Draft: Provide initialization documentation for cloning Dandi ecosystem (#104)
dandi/handbook: docs/40_initialization.md

<br/><br/>

Lastly, add them to the Group you made in the step above:

<br/><br/>
<img
src="../img/add_perms.png"
alt="add_perms"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

##### Create security credentials

You'll need to create `Security Credentials` for your User -- these credentials will be specifically used
in your `Terraform Cloud` setup

Firstly, go to your User and click the `Security Credentials` tab:

<br/><br/>
<img
src="../img/creds.png"
alt="creds"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>

Navigate to `Create access key`

<br/><br/>
<img
src="../img/access_key.png"
alt="access_key"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

You'll be prompted to provide a reason for the access key creation. It doesn't matter much, so `Other` is a completely acceptable choice

<br/><br/>
<img
src="../img/creds_purpose.png"
alt="creds_purpose"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Time to create the key!

<br/><br/>
<img

> src="../img/create_tag.png"

Reviewer: that is a lot of pngs... all adds up - would make git repo heavy etc.  
ATM  it is not that much -- almost 5MB
```
❯ git diff master... --stat | grep '\.png' | awk '{print $1;}' | xargs du -sck | grep total
4484	total
```
but if converted to .jpg and on such large ones sacrificing some quality -- we can get them 3 times smaller and only with minimal amount of distortions IMHO tollerable for this tech doc:
```
❯ git diff master... --stat | grep '\.png' | awk '{print $1;}' | while read f; do convert $f -quality 30 ${f//.png/.jpg}; done
❯ git diff master... --stat | grep '\.png' | sed -e 's,.png,.jpg,g' | awk '{print $1;}' | xargs eog
❯ git diff master... --stat | grep '\.png' | sed -e 's,.png,.jpg,g' | awk '{print $1;}' | xargs du -sck | grep total
1484	total
```
WDYT?

Reviewer 2: Thanks Yarik.  Good catch.

> # Getting Started
> 
> This series of docs displays how to create a clone of DANDI. It is suggested that you briefly read through each of the
> documents in this guide before starting.

Reviewer 2: ```suggestion
# Initialize Vendor Accounts

This series of documents explains how to create a new instance of the DANDI project (i.e. a clone of the DANDI Archive). It is suggested that you briefly read through each of the
documents in this guide before starting.
```

> Other than cloning repositories and writing code, you'll first need to set up a handful of accounts. These accounts help facilitate
> the creation of services and libraries that eventually can be available to the public.
> 
> The Dandi ecosystem relies on a handful of vendor services to operate:
> 
> The following accounts must be made to create your Dandi clone in a production setting:

Reviewer 2: ```suggestion
The DANDI ecosystem relies on vendor services to operate.  So first you will need to set up accounts with the following vendors:
```

<br/><br/>
<img
src="../img/creds.png"
alt="creds"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>

Navigate to `Create access key`

<br/><br/>
<img
src="../img/access_key.png"
alt="access_key"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

You'll be prompted to provide a reason for the access key creation. It doesn't matter much, so `Other` is a completely acceptable choice

<br/><br/>
<img
src="../img/creds_purpose.png"
alt="creds_purpose"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Time to create the key!

<br/><br/>
<img
src="../img/create_tag.png"
alt="create_tag"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

You'll be provided with the values of your `Access key` and `Secret access key` -- store these values somewhere secure and accessible

<br/><br/>
<img
src="../img/retrieve.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

## GitHub

> You'll need to create a GitHub Organization with your DANDI fork. [See here for documentation to create a GitHub organization](https://docs.github.com/en/organizations/collaborating-with-groups-in-organizations/creating-a-new-organization-from-scratch)

Reviewer 2: ```suggestion
You'll need to create a GitHub Organization where you can fork the DANDI repositories. [See here for documentation to create a GitHub organization](https://docs.github.com/en/organizations/collaborating-with-groups-in-organizations/creating-a-new-organization-from-scratch)
```

<br/><br/>
<img
src="../img/secrets_actions.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

And then lastly create two new secrets, `HEROKU_EMAIL` should be the email address/account you used to generate your `HEROKU_API_KEY`

<br/><br/>
<img
src="../img/heroku_secrets.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

## Terraform Cloud

Terraform is configuration tool for managing "infrastructure-as-code" -- meaning that we can programatically manage infrastructure in
a traceable, version-controlled form.

The Terraform ecosystem provides a UI tool called **Terraform Cloud**

Start by visiting `https://app.terraform.io/` and making an account.

##### Creating a Terraform Project and Workspace

Once you have successfully made an account, you'll want to create a `Workspace` and a `Project` in that `Workspace`

<br/><br/>
<img
src="../img/terraform_init.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

##### Populate Environment Variables

> After successful creation, you'll want to populate the following variables in the `Variables` section of the `Workspace`:
> 
> You'll need to populate `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `HEROKU_API_KEY` and `HEROKU_EMAIL`

Reviewer 2: Thanks Aaron.  To simplify, I would suggest removing the phrase `you'll want to` from the instructions.

```suggestion
Populate the following variables in the `Variables` section of the `Workspace`:

1. `AWS_ACCESS_KEY_ID`
2. `AWS_SECRET_ACCESS_KEY`
3. `HEROKU_API_KEY`
4. `HEROKU_EMAIL`
```

<br/><br/>

You'll want to next connect `dandi-archive` via `Deploy with GitHub`. Enable Netlify to be authorized as a GitHub app. Once you have
enabled authorization, you'll need to specifically link your appropriate repositories:

<br/><br/>
<img
src="../img/netlify_auth.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

Now once you see the appropriate repository, you'll want to navigate to configure where Netlify can find and build your site

<br/><br/>
<img
src="../img/netlify_repo_search.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

For usage of Netlify, one could refer to declaring a `netlify.toml` configuration file [like the one referenced in DANDI Archive](https://github.com/dandi/dandi-archive/blob/master/web/netlify.toml)

These values can also be replicated in the settings. 

<br/><br/>
<img
src="../img/netlify_config.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

> Your frontend should be able to deploy to a auto-generated URL via Netlify now! Steps for domain management and configuration are described further in the [Frontend Deployment](.
> ./42_frontend_deployment) section of these docs.

Reviewer 2: ```suggestion
Your frontend should be able to deploy to an auto-generated URL via Netlify now! Steps for domain management and configuration are described further in the [Frontend Deployment](.
./42_frontend_deployment) section of these docs.
```

## Sentry

[Sentry](https://sentry.io/) is a monitoring tool used for the DANDI Archive API. It is integral in order to notify engineers if a system is down, experiencing poor performance, or may have unwanted users

Begin by creating a Sentry account -- once successful, you'll start by creating a new Project:

<br/><br/>
<img
src="../img/sentry_init.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

##### Select Django as an App Type

DANDI Archive API is built as a Django app -- so proceed to select `Django` on the following screen:

<br/><br/>
<img
src="../img/sentry_django.png"
alt="retrieve"
style="width: 60%; height: auto; display: block; margin-left: auto;  margin-right: auto;"/>
<br/><br/>

##### Capture Sentry DSN value

You'll be provided with a screen displaying how to initialize and install Sentry into your Django app. For now, just capture the DSN value.

> This value will be used later as an environment variables while deploying your API via Terraform

Reviewer 2: ```suggestion
This value will be used later as an environment variable while deploying your API via Terraform.
```
